src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -95,47 +105,77 @@ public void visitToken(DetailAST ast) {
-                    if (modifiers.branchContains(TokenTypes.LITERAL_PRIVATE)) {
-                        desc.registerPrivateCtor();
-                    }
-                    else {
-                        desc.registerNonPrivateCtor();
-                    }
-                }
+                registerClassConstructors(ast, modifiers);
                 break;
 
             default:
                 throw new IllegalStateException(ast.toString());
         }
     }
 
+    /**
+     * Register class constructors.
+     *
+     * @param ast       - AST node
+     * @param modifiers - node with first modifiers token
+     */
+    private void registerClassConstructors(DetailAST ast, DetailAST modifiers) {
+        if (!ScopeUtils.isInEnumBlock(ast)) {
+            final ClassDesc desc = classes.peek();
+            if (modifiers.branchContains(TokenTypes.LITERAL_PRIVATE)) {
+                desc.registerPrivateCtor();
+            } else {
+                desc.registerNonPrivateCtor();
+            }
+        }
+    }
+
+    /**
+     * Build class description from ast node.
+     *
+     * @param ast       - AST node
+     * @param modifiers - node with first modifiers token
+     * @return Class description
+     */
+    private ClassDesc buildClassDesc(DetailAST ast, DetailAST modifiers) {
+        final boolean isFinal = modifiers.branchContains(TokenTypes.FINAL);
+        final boolean isAbstract = modifiers.branchContains(TokenTypes.ABSTRACT);
+        final String superClass = getSuperClassName(ast);
+        if (superClass != null) {
+            superClasses.add(superClass);
+        }
+        final String qualifiedClassName = getQualifiedClassName(ast);
+        final ClassDesc classDesc = new ClassDesc(qualifiedClassName, isFinal, isAbstract);
+        classDesc.astNode = ast;
+        classDesc.depthLevel = currentDepthLevel;
+        return classDesc;
+    }
+
     @Override
     public void leaveToken(DetailAST ast) {
         if (ast.getType() == TokenTypes.CLASS_DEF) {
-            final ClassDesc desc = classes.pop();
-            if (desc.isWithPrivateCtor()
-                && !desc.isDeclaredAsAbstract()
-                && !desc.isDeclaredAsFinal()
-                && !desc.isWithNonPrivateCtor()
-                && !desc.isWithNestedSubclass()
-                && !ScopeUtils.isInInterfaceOrAnnotationBlock(ast)) {
+            currentDepthLevel--;
+        }
+    }
+
+    @Override
+    public void finishTree(DetailAST rootAST) {
+        determineParentClasses();
+        for (ClassDesc desc : classes) {
+            if (shouldDeclaredAsFinal(desc)) {
                 final String qualifiedName = desc.getQualifiedName();
                 final String className = getClassNameFromQualifiedName(qualifiedName);
-                log(ast.getLineNo(), MSG_KEY, className);
+                log(desc.astNode.getLineNo(), MSG_KEY, className);
             }
         }
+        classes.clear();
+        superClasses.clear();
```

Comment:
These 2 clears don't seem needed. We erase these classes on brand new files in `beginTree`.